src/expr/src/expr/array.rs

```diff
@@ -0,0 +1,214 @@
+/// ```
+///
+/// Examples:
+///
+/// ```slt
+/// query T
+/// select array_to_string(array[1, 2, 3, NULL, 5], ',')
+/// ----
+/// 1,2,3,5
+///
+/// query T
+/// select array_to_string(array[1, 2, 3, NULL, 5], ',', '*')
+/// ----
+/// 1,2,3,*,5
+///
+/// query T
+/// select array_to_string(array[null,'foo',null], ',', '*');
+/// ----
+/// *,foo,*
+///
+/// query T
+/// with t as (
+///   select array[1,null,2,3] as arr, ',' as d union all
+///   select array[4,5,6,null,7] as arr, '|')
+/// select array_to_string(arr, d) from t;
+/// ----
+/// 1,2,3
+/// 4|5|6|7
+///
+/// # `array` or `delimiter` are required. Otherwise, returns null.
+/// query T
+/// select array_to_string(array[1,2], NULL);
+/// ----
+/// NULL
+///
+/// # FIXME: this query will panic in binder
+/// # query T
+/// # select array_to_string(NULL, ',');
+/// # ----
+/// # NULL
+/// ```
+#[derive(Debug)]
+pub struct ArrayToStringExpression {
+    pub array: Box<dyn Expression>,
+    pub delimiter: Box<dyn Expression>,
+    pub null_string: Option<Box<dyn Expression>>,
+}
+
+impl<'a> TryFrom<&'a ExprNode> for ArrayToStringExpression {
+    type Error = ExprError;
+
+    fn try_from(prost: &'a ExprNode) -> Result<Self> {
+        ensure!(prost.get_expr_type().unwrap() == Type::ArrayToString);
+        let RexNode::FuncCall(func_call_node) = prost.get_rex_node().unwrap() else {
+            bail!("Expected RexNode::FuncCall");
+        };
+        let mut children = func_call_node.children.iter();
+        let Some(array_node) = children.next() else {
+            bail!("Expected argument 'array'");
+        };
+        let array = build_from_prost(array_node)?;
+
+        let Some(delim_node) = children.next() else {
+            bail!("Expected argument 'delimiter'");
+        };
+        let delimiter = build_from_prost(delim_node)?;
+
+        let null_string = if let Some(null_string_node) = children.next() {
+            Some(build_from_prost(null_string_node)?)
+        } else {
+            None
+        };
+
+        Ok(Self {
+            array,
+            delimiter,
+            null_string,
+        })
+    }
+}
+
+impl Expression for ArrayToStringExpression {
+    fn return_type(&self) -> DataType {
+        DataType::Varchar
+    }
+
+    fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let list_array = self.array.eval_checked(input)?;
+        let list_array = list_array.as_list();
+
+        let delim_array = self.delimiter.eval_checked(input)?;
+        let delim_array = delim_array.as_utf8();
+
+        let null_string_array = if let Some(expr) = &self.null_string {
+            let null_string_array = expr.eval_checked(input)?;
+            Some(null_string_array)
+        } else {
+            None
+        };
+
+        let mut output = Utf8ArrayBuilder::with_meta(input.capacity(), ArrayMeta::Simple);
+
+        for (i, vis) in input.vis().iter().enumerate() {
+            if !vis {
+                output.append_null();
```

Comment:
nit: invisible and array/delim null `append_null` could use `continue` to avoid deep indentation.